File: app/src/main/java/com/nimtego/plectrum/data/network/itunes/ITunesApi.java
package com.nimtego.plectrum.data.network.itunes;

import com.nimtego.plectrum.data.model.itunes.AlbumsRepository;
import com.nimtego.plectrum.data.model.itunes.ArtistsRepository;
import com.nimtego.plectrum.data.model.itunes.SongsRepository;

import java.util.Map;

import io.reactivex.Observable;
import retrofit2.http.GET;
import retrofit2.http.QueryMap;

public interface ITunesApi {

    @GET("/search")
    Observable<ArtistsRepository> searchArtist(@QueryMap Map<String, String> param);

    @GET("/search")
    Observable<AlbumsRepository> searchAlbum(@QueryMap Map<String, String> param);

    @GET("/search")
    Observable<SongsRepository> searchSongs(@QueryMap Map<String, String> param);

    @GET("/lookup")
    Observable<SongsRepository> getSongs(@QueryMap Map<String, String> param);

    @GET("/lookup")
    Observable<ArtistsRepository> getArtist(@QueryMap Map<String, String> param);

    @GET("/lookup")
    Observable<AlbumsRepository> getAlbum(@QueryMap Map<String, String> param);
}

File: app/src/main/java/com/nimtego/plectrum/data/network/itunes/ITunesApi.kt
package com.nimtego.plectrum.data.network.itunes

import com.nimtego.plectrum.data.model.itunes.AlbumsResultRepository
import com.nimtego.plectrum.data.model.itunes.ArtistsResultRepository
import com.nimtego.plectrum.data.model.itunes.MovieResultRepository
import com.nimtego.plectrum.data.model.itunes.SongsResultRepository

import io.reactivex.Observable
import retrofit2.http.GET
import retrofit2.http.QueryMap

interface ITunesApi {

    @GET("/search")
    fun searchArtist(@QueryMap param: Map<String, String>): Observable<ArtistsResultRepository>

    @GET("/search")
    fun searchAlbum(@QueryMap param: Map<String, String>): Observable<AlbumsResultRepository>

    @GET("/search")
    fun searchSongs(@QueryMap param: Map<String, String>): Observable<SongsResultRepository>

    @GET("/search")
    fun searchMovie(@QueryMap param: Map<String, String>): Observable<MovieResultRepository>

    @GET("/lookup")
    fun getMovie(@QueryMap param: Map<String, String>): Observable<MovieResultRepository>

    @GET("/lookup")
    fun getSongs(@QueryMap param: Map<String, String>): Observable<SongsResultRepository>

    @GET("/lookup")
    fun getArtist(@QueryMap param: Map<String, String>): Observable<ArtistsResultRepository>

    @GET("/lookup")
    fun getAlbum(@QueryMap param: Map<String, String>): Observable<AlbumsResultRepository>


}
